### PR TITLE
Enhancing the dav pluginmanager

### PR DESCRIPTION
**Pull request by rullzer**
Psalm was complaining a bit. But when going over the code there was a lot of magic going on under the hood.
This is an attempt at restructuring this a bit.

### CODE REVIEW

**Review by ChristophWurst — APPROVED** [review]
Lots of new warnings :P

**Comment by rullzer** [comment]
> Lots of new warnings :P

Those warnings were already there. It is beause we return mixed from the appmanager. (note how it is always the same code it is complaining about).
The rest of the class is cleaner now.

File: apps/dav/lib/AppInfo/PluginManager.php [hunk 1]
```diff
@@ -265,42 +260,64 @@ private function createPluginInstance(string $className) {
...
+	 * @return ICalendarProvider[]
+	 */
+	private function loadSabreCalendarPluginsFromInfoXml(array $classes): array {
+		return array_map(function (string $className): ICalendarProvider {
+			$instance = $this->createClass($className);
+			if (!($instance instanceof ICalendarProvider)) {
+				throw new \Exception('Sabre calendar plugin class ' . $className . ' does not implement the ' . ICalendarProvider::class . ' interface');
+			}
+			return $instance;
+		}, $classes);
```
Comment (rullzer) [inline apps/dav/lib/AppInfo/PluginManager.php:321]: Like I'm sure we could also make a generic function that does this. But currently we have 4 of them. So I don't mind the duplication.

File: apps/dav/lib/AppInfo/PluginManager.php [hunk 2]
```diff
@@ -111,56 +114,70 @@ public function getAppPlugins() {
...
 			// load plugins and collections from info.xml
 			$info = $this->appManager->getAppInfo($app);
 			if (!isset($info['types']) || !in_array('dav', $info['types'], true)) {
 				continue;
 			}
-			$this->loadSabrePluginsFromInfoXml($this->extractPluginList($info));
-			$this->loadSabreCollectionsFromInfoXml($this->extractCollectionList($info));
-			$this->loadSabreAddressBookPluginsFromInfoXml($this->extractAddressBookPluginList($info));
-			$this->loadSabreCalendarPluginsFromInfoXml($this->extractCalendarPluginList($info));
+			$plugins = $this->loadSabrePluginsFromInfoXml($this->extractPluginList($info));
```
Comment (kesselb) [inline apps/dav/lib/AppInfo/PluginManager.php:154]:
```suggestion
			$this->plugins = $this->loadSabrePluginsFromInfoXml($this->extractPluginList($info));
```

maybe? I don't see a second write to $this->plugins.

**Review by kesselb — APPROVED** [review]

File: apps/dav/lib/AppInfo/PluginManager.php (hunk 2 above)
Comment (rullzer) [reply apps/dav/lib/AppInfo/PluginManager.php:154]: It loops ;) so that would overwite it

**Pull request merged**